### ISSUE TITLE
File Transport Checksums Between Server and Desktop

### ISSUE THREAD

**Issue by MTRichards** [owncloud/client#3735]
As an ownCloud admin, I want ownCloud to ensure that files transported from the desktop client to the server are not corrupted in transit so that we can catch bugs and avoid corrupting customer data

Acceptance Criteria:
- Desktop sync client takes a checksum of files on the desktop when they are discovered and stores the checksum in the database
- When the file is transmitted to the ownCloud server, the checksum is also transmitted
- The server checksums the file when it arrives at the server, compares it to the checksum transmitted by the client previously
- If the checksums match, a success return code is sent
- If the chechsums fail to match, the file has been corrupted and the desktop skips that file. The file will be retransmitted on the next upload.
- If the file is corrupted in transit, the file transfer log on the desktop reflects that this file was corrupted in transit, and that it will be retried on the next sync run 
Note: this links to server issue XXX

**Comment by ckamm** [comment]
What needs to be done in the client?
* Determine whether the server accepts checksums, possibly via capabilities. Currently checksumming needs to be explicitly enabled in the config file.
* Store checksums in the database. (for performance optimization only, I guess?)
* Check that checksum errors are handled nicely.

**Comment by ckamm** [comment]
We also need to make sure getting an empty checksum header from the server during downloads isn't considered a validation failure - currently that seems to be the case. I wonder why b05ca526a46c8a306ccfb4aca0d5cf94881ba14f did that -- @dragotin?

**Comment by dragotin** [comment]
@ckamm the patch you're pointing to checks the expected checksum **type** as in Adler32 or SHA or so. If that is empty, we do not have checksumming and assume validated() - which means all is ok.

**Comment by ckamm** [comment]
@dragotin Yes, but why would the checksum type configured in the config file matter when one validates a checksum the server sends? 

And shouldn't an empty checksum header still yield `validated`? It sounds like only files that were uploaded by new clients will have checksums when using the owncloud server.

**Comment by ckamm** [comment]
@moscicki I'm touching the checksumming code and wonder: Do you use `Adler32` or `ADLER32`? What does the server accept and what does it send? I'd prefer to just use one spelling and there's a small chance you're requiring one for uploads and the other for downloads...

**Comment by moscicki** [comment]
We treat the request header name as case-insensitive. So all these are accepted: Adler32, adler32, ADLER32

Response given by the server: Adler32

**Comment by ckamm** [comment]
@moscicki Great, then the client will standardize on `Adler32`.

**Comment by ogoffart** [comment]
The problem with Adler32 is that it is not safe against collision. And so we cannot use it to detect copies or moves. (Or can we?)

**Comment by dragotin** [comment]
@moscicki can you please enlighten us?

**Comment by moscicki** [comment]
What exactly do you want to use the checksum for? As Etag? Or deduplicate
the transfer in case of a copy or simple "touch"? Not sure I understand how
moves can be helped with the checksum...?

I think the answer depends on what you want to do with it ;-)


On Wed, Oct 21, 2015 at 6:24 PM, Klaas Freitag <notifications@github.com>
wrote:

> @moscicki <https://github.com/moscicki> can you please enlighten us?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/3735#issuecomment-149951450>.
>



-- 
---
Best regards,
Kuba

**Comment by ogoffart** [comment]
This is not decided yet. But if we have checksum in the database we might as well use it to detect copies and move more reliably than now.

**Comment by moscicki** [comment]
I think this requires an architectural discussion, also in the context of new storage architecture promised for owncloud 9 server. This is to say that IMO sync client should be flexible enough to handle storage backends which come with checksums already (e.g. S3 with MD5, EOS with Alder32,...). This is the storage backend behind the owncloud server which tells the client what the checksum is - you cannot change that unless you want to compute it yourself again in the owncloud server. For file downloads the checksum type comes in the response header -- so client just uses this information. For file uploads client may want to discover what is the checksum type supported by the server (this may be per-directory!) in order to compute the expected checksum and add it to the PUT header. Except for the discovery this is what has already been implemented in the sync client and for protecting the file transfer itself this should be sufficient. 

Another point is what kind of checksum you store locally in the sync state db and if this could/should be different from the transmission checksum (at the expense of computing the checksum twice) and how that can help you to optimize operations. If this can help you at all then it should be a very strong checksum that would guarantee that if checksum1==checksum2 is equivalent to content1=content2. This is not true for Adler32. Here are some pointers: 

 - https://cds.cern.ch/record/2026187/files/Adler32_Data_Corruption.pdf
 - http://www.leviathansecurity.com/blog/analysis-of-adler32

Having said that, it probably makes sense to first try to understand what kind of optimizations you may want to achieve (deduplicating transfer in case of propagating a local copy -- yes, that would be one such case -- but is it worth it?). For moves, I really do not see how this can help the client to track them.

And, in addition to all that, checksuming should be considered in the context of delta sync -- this is potentially a big win. But this needs to be verified on real data.

The bottom line is that one should analyze the real usage of the system to have some idea about how important these things are. We do this kind of data mining on CERNBox currently so any ideas about what pieces of information to extract to provide a useful, real-life information as input to this discussion are welcome.

I propose to prepare the grounds and aim at the discussion in Zurich CS3 Workshop in ETH in January.

**Comment by dragotin** [comment]
Yes, actually there are three areas where we can use checksums, and often these are confused in discussions:

### Checksums for Transportation Verification (#3735)

The idea here is to verify the integrity of a file's content after it has been transmitted over any kind of network, usually after it has been downloaded from the server or uploaded from a client to the server. A checksum is added to the header of the GET or PUT request and the receiving site recalculates the content checksum of the file on it's final location and compares it to the expected value from the header. This adds another level of safety to the transmission.

For this usecase, any available checksum from the storage backends should be used, to keep the effort for the server as low as possible. The header that contains the checksum information for this reason prefixes the checksum with its type, for example `SHA1:543353xadr32323`.

If the type is unkown to the receiving site, the transmission verification can not be done. This is currently not considered an error, but on the longer run, it could become a configuration option if it should.

### Checksums to Support Discovering Changes

This usecase has two aspects:

1. Verification of moves: Currently moves are detected by comparison of file size, it's inode and modification time. All values have to be equal. Still there could be cases where this is not a move but a new file, so verification of the content checksum which also must be equal would solve that. (See #2983)
2. Detection of "wrong" change notification on local file system (the .eml problem, #3235): Even if the modification time changed, a content checksum could be used in this case to check if it really was a change or not.
3. Avoid *touch* changes: Currently ownCloud syncs files that do not have a content change, but only a change of the modification time, ie. through command *touch*. This could be avoided if a content checksum check is performed if the modification time changed, but not the file size.

These usecase is mainly isolated on a single site, so any checksum that is available and stored in the journal could be used.
 
### Checksums to Enable Delta Sync

The main difference of this compared to the two other usecases is that here checksums over blocks of a file are needed, and not checksums over the entire file. The usual delta sync algorithms usually split a file into blocks of equal size, calculate the checksums over these blocks and try to only transmit these blocks that have really changed. The tricky part is how the file is split to have a maximum high probability of finding unchanged blocks. (See #179)

For this usecase, strong checksums are needed. Both client and server need to be able to recalculate and possibly store the checksum lists.

**Comment by moscicki** [comment]
That's a very nice summary! I add this link for future reference: 

http://stackoverflow.com/questions/12186993/what-is-the-algorithm-to-compute-the-amazon-s3-etag-for-a-file-larger-than-5gb

**Comment by ckamm** [comment]
@Dianafg76 Since the server doesn't yet support this, testing this is a bit involved:
* Set "transmissionChecksum=MD5" in owncloud.cfg, in the "[General]" section
* Set up the client to go through mitmproxy
* Look at the PUT requests in mitmproxy, they should contain a OC-Checksum header
* You can use md5sum on the file to check the checksum correctness
* To check that the uploaded checksums also end up in the database, look at the transmissionChecksum column in the metadata table with a tool like sqlitebrowser
* To check that the checksum in GET responses is handled correctly, configure a Header-rewrite rule in mitmproxy. My filter was "~s" "OC-Checksum" "MD5:aabbeeff" - you can put the correct or an incorrect checksum there for testing.

Since this is exactly what I did for testing, it may not make much sense for you to do this again. Maybe we should defer testing until the server actually supports the checksums.

@moscicki It'd be awesome if you could test the client against your backend.

**Comment by Dianafg76** [comment]
@ckamm OK, thanks

**Comment by moscicki** [comment]
Which version would you like me to test? We just tested 2.0.2 and checksum headers are present.

Since we also recently set up mitmproxy but we do not have experience yet, do you have any internal doc how to configure it that could help us? BTW. Is there a way of automating the mitmproxy testing in some way? Something like controlling the mitmproxy via a CLI to enable disable checksum corruption?

If you want to test how the server responds to the incorrect checksum headers, you may use this test automatically: https://github.com/cernbox/smashbox/blob/master/protocol/test_protocol_file_checksum.py

**Comment by ckamm** [comment]
@moscicki Significant improvements to the checksum handling just went into master, which will become 2.1. Tomorrow's nightly builds for that version will include the changes.

mitmproxy: No, we don't have extra documentation. I basically just start it and manually configure it as the proxy in the client and check the data flow. I haven't looked at automating tests using it yet. It does have python scripting capability and a non-interactive companion utility (mitmdump) that could presumably be used though.

**Comment by moscicki** [comment]
Hello,

For the moment I notice a problem with 2.0.2 against our server (for which checksuming in 1.7.2 works fine):

The checksum header is malformed. See below.

![image](https://cloud.githubusercontent.com/assets/621965/10908416/29fc5f34-8232-11e5-9132-d9ad00206728.png)

**Comment by dragotin** [comment]
@moscicki is that happening for upload only? And can you see how the checksum header actually looks like?

**Comment by dragotin** [comment]
@moscicki which  platform is that? 

My guess is that zlib is missing for https://github.com/owncloud/client/blob/master/src/libsync/transmissionchecksumvalidator.cpp#L92 and thus no Adler32 support is compiled in.

**Comment by dragotin** [comment]
@michaelstingl FYI

**Comment by moscicki** [comment]
It is MacOSX.

**Comment by michaelstingl** [comment]
@dragotin Is there anything I can do?

**Comment by michaelstingl** [comment]
@dragotin What is the status here? /cc @MorrisJobke 

00004230

**Comment by MorrisJobke** [comment]
> @dragotin What is the status here? /cc @MorrisJobke

Missing implementation on server side. Therefore the client implementation also needs to wait.

**Comment by ckamm** [comment]
@moscicki "The checksum header is malformed." should only happen when the header is non-empty but does not contain a colon, since the expected format is `type:checksum`. If you start it from a console you should be seeing a `Checksum header malformed: <header-val>` qDebug message: what checksum header does the server send?

**Comment by dragotin** [comment]
The build environments have been checked and fixed in regards of availability of zlib.

**Comment by dragotin** [comment]
@labkode do you have a chance to verify this issue? It seems @moscicki is out for a couple of days, but we would love to have that clearified before we release 2.1.0

**Comment by labkode** [comment]
@dragotin I don't have access to CERN EOS right now but I'll contact someone to test that ASAP.

Meanwhile you can use ClawIO, it supports the OwnCloud Sync Protocol with checksums.

After testing against ClawIO I think there are two problems that lead to these:

1) Server does not send the OC-Checksum because the file has not checksum (so we avoid sending an empty header)

2) Case problem: if client sends MD5:xxx but server replies with md5:xxx then the error is triggered.

In my opinion I think that the the client should be case-insensitive with the checksum payload, it is easier for everybody. 

If you want to try out your client checksumming the fastest way is to use ClawIO.

# ClawIO Setup

You need to have `docker` and `docker-compose` in your machine.

Then run this script to set up the ClawIO server:

````
#!/usr/bin/bash

mkdir -p /tmp/clawioproject
cd /tmp/clawioproject/

git clone https://github.com/clawio/orches
git clone https://github.com/clawio/service.auth
git clone https://github.com/clawio/service.localstore.data
git clone https://github.com/clawio/service.localstore.meta
git clone https://github.com/clawio/service.ocwebdav
git clone https://github.com/clawio/service.localstore.prop

cp orches/docker-compose.yml .

docker-compose up -d --force-recreate
docker-compose up -d
docker-compose up -d
docker-compose up -d
docker-compose up -d

docker-compose ps
```

Before using the OwnCloud client you have to use the ClawIO CLI to setup your user:

```
 docker run -it --link "service.auth:service-auth" --link "service.localstore.data:service-localstore-data" --link "service.localstore.meta:service-localstore-meta" labkode/clawio bash
```

Once inside the container run:

```
clawio login demo demo
clawio home
clawio stat /local/users/d/demo --children
```

You can upload some files and create some directories:

```
clawio mkdir /local/users/d/demo/folderone
clawio upload /etc/passwd /local/users/d/demo/file
clawio upload /etc/passwd /local/users/d/demo/filewithchecksum --checksum="md5:$(md5sum /etc/passwd | cut -d ' ' -f1)"
clawio stat /local/users/d/demo --children
```

## OwnCloud Client with ClawIO Server
The ClawIO OwnCloud server listens on port 57004.
If you run docker on your machine the connection should be `http://localhost:57004`.
If you are in a Mac, the default ip is `docker-machine ip default`

The user and password is `demo:demo`

You have to click `skip folder configuration`.

Then you click on Add folder and the remote folder is `/local/users/d/demo`.

The sync client can sync now.

To see the logs on the ClawIO server: 

````
cd /tmp/clawioproject
docker-compose logs
```

Sometimes the server fails to upload due to docker filesystem and you will see EOF errors.


Hope it helps ;)

**Comment by labkode** [comment]
## Tests performed against ClawIO

If the server responds with **lowercase checksum type** (md5 instead of MD5) the sync client reports with the malformed error **BUT the file APPEAR it in the local folder.**

If server does **NOT send the header** the client reports the same error **BUT the file does NOT APPEAR in the local folder**

Example response with checksum:

```
Content-Type: application/octet-stream
Etag: 31b699cf-b3c9-488a-9fed-8fec2f5c4873
Last-Modified: Wed, 25 Nov 2015 21:32:57 CET
Oc-Checksum: md5:c0d388ca626db569e13fdee7ad41c9bf
Oc-Etag: 31b699cf-b3c9-488a-9fed-8fec2f5c4873
Oc-Fileid: 2c682baa-7c17-498f-a05d-a100762f3e08
Date: Wed, 25 Nov 2015 21:25:36 GMT
Content-Length: 1197
```

**Comment by ckamm** [comment]
@labkode Thanks for testing!

1) Which client version were you testing with? The client behavior you report surprises me and I'd guess you're using an older version?

2) No header in downloads: If no header is sent, the client should not report an error and should just save the file. This is essential behavior that guarantees compatibility with servers that don't send transmission checksums and almost certainly works. This is why I wonder what client version you're testing.

3) Checksum type strings: This is a real problem. We could make all checksum strings case-insensitive in the client. The client will still send a specific string (currently MD5, SHA1 or Adler32) to the server for uploads and the server would need to accept that.

**Comment by labkode** [comment]
>1) Which client version were you testing with? The client behavior you report surprises me and I'd guess >you're using an older version?

OwnCloud Version 2.0.2 (build 2818)

>2) No header in downloads: If no header is sent, the client should not report an error and should just >save the file. This is essential behavior that guarantees compatibility with servers that don't send >transmission checksums and almost certainly works. This is why I wonder what client version you're >testing.
OwnCloud Version 2.0.2 (build 2818)

>3) Checksum type strings: This is a real problem. We could make all checksum strings case-insensitive >in the client. The client will still send a specific string (currently MD5, SHA1 or Adler32) to the server for >uploads and the server would need to accept that

Yes, both client and server should be case-insensitive for just the checksum type, the checksum value should be case sensitive.


CERNBox client 2.0.2 is also reporting the malformed errors.

 We are going test both OwnCloud 2.0.2 and CERNBox 2.0.2 against EOS and get you feedback ASAP.

**Comment by ckamm** [comment]
@labkode This is about changes to the code that were made for OwnCloud client 2.1.0. Please test http://download.owncloud.com/desktop/daily/ or https://build.opensuse.org/package/show/isv:ownCloud:community:nightly/owncloud-client . We're also mainly interested in the 'Adler32' checksum mode, because that's what is used in practice.

I'm in the process of setting up a test env for myself too.

**Comment by ckamm** [comment]
@labkode I had some trouble setting up the clawio environment. It seems all the links between the containers failed to resolve (e.g. service.localstore.prop failed to look up service-localstore-prop-mysql). The error was similar to https://github.com/gogits/gogs/issues/1784. So I fixed by using the recommended 'localhost:...' addresses instead of the linked service names.

Afterwards I could successfully sync files when Adler32 transmission checksums where enabled in the client. Using MD5 as a transmission checksum also worked correctly.

**Comment by ckamm** [comment]
@labkode Okay, now that we know that the tests against clawio pass, I think the main thing remaining is to test against EOS. For that, you could use the 2.1.0RC1 version which is going to be out soon. (Please set `transmissionChecksum=Adler32` in owncloud.cfg, in the `[General]` section)

**Comment by labkode** [comment]
@ckamm Great!!!

I also tested again ClawIO with OwnCloud 2.1.0-nightly20151125 (build 2907) and now the behaviour is correct.

To sum up:

```
Files without checksum appear in local folder
Files with a checksum type different from the one specified by the client appear in the local folder
Files with the wrong-case checksum type do not appear and client complains with malformed ...
```

Yes, I'll test against EOS as soon as I get access to it and the RC1 is released.

**Comment by labkode** [comment]
@ckamm @dragotin 
Tested with OwnCloud 2.1.0-nightly20151125 (build 2907) against our EOS instance and the client behaves correctly when Adler32 is used.

Waiting to test RC1

**Comment by dragotin** [comment]
@labkode did you have a chance to test RC1?

**Comment by labkode** [comment]
@dragotin

I couldn't find the release on the Mac nightlies. Can you point me to the RC?

**Comment by dragotin** [comment]
It's https://download.owncloud.com/desktop/testing/ownCloud-2.1.0.2914rc1.pkg via 
https://owncloud.org/install/#testing-development
Thanks!

**Comment by labkode** [comment]
@dragotin 

The client behaves correctly.

But the problem with the case still persists, but with a different error.

```
01/12/15 11:00:31fromweb2.txt                                                    ownCloud                      The checksum header contained an unknown checksum type 'ADLER32'
01/12/15 10:59:54config.toml                                                     ownCloud                      The checksum header contained an unknown checksum type 'ADLER32'
```

**Comment by dragotin** [comment]
@labkode sorry, can you elaborate? Is that a problem with EOS? Or with the configuration in the config file? Where do you see the error?

**Comment by labkode** [comment]
@dragotin 

As explained in comments before, if the server checksum type has a different case that the one configured in the client (Adler32 vs ADLER32 or MD5 vs md5) the client do not accept the checksum.

With EOS is working correctly. EOS sends Adler32 in the response and the client is configured with Adler32. No problem.

To trigger this error I hijacked the response from EOS to insert ADLER32 checksum type.
The error appears in the client log when the client tries to download the file.

Is the RC1 client supposed to fix the case problem ?

Cheers

**Comment by dragotin** [comment]
Thanks @labkode 

I am not sure if this is really an error. I mean, md5 != MD5, so I think this is an acceptable behaviour. This feature **needs** attention from the administrator, so configuring it correctly should be possible.

So we can consider this working I assume?

**Comment by labkode** [comment]
@dragotin The client works correctly on Mac OSX. I have to try on Linux and Windows. But if the codebase is the same they should also work correctly.

**Comment by labkode** [comment]
@dragotin  Tested from Win and Linux. Both working against EOS with Adler32.

**Comment by dragotin** [comment]
Reopening because we have to test that again against ownCloud server.

**Comment by ckamm** [comment]
@bboule Are you saying you're currently working on this? Should I assign this to you?

**Comment by MorrisJobke** [comment]
> @bboule Are you saying you're currently working on this? Should I assign this to you?

I don't think so. This is just a label of the tool he is using :see_no_evil:

**Comment by ckamm** [comment]
@MorrisJobke Thanks, that explains it. As far as I know, testing this can only start when owncloud/core#18716 is done.

**Comment by rullzer** [comment]
See https://github.com/owncloud/core/issues/21997 for a WIP PR... it can store provided checksums from the OC-Checksum Header. Will return the OC-Checksum header if there is a checksum. And on propfind the checksums will be returned as well `<oc:checksums>`

The format for the header field is basically: `<TYPE>:<CHECKSUM>(, <TYPE>:<CHECKSUM>)*`

**Comment by rullzer** [comment]
https://github.com/owncloud/core/pull/21997 is in. However slimmed down. So right now only 1 checksum can be send.

But all should work for 9.0

**Comment by dragotin** [comment]
The acceptance criteria of this issue is not met as the server in the current implementation (9.0) does not validate the checksum that comes with the client's file upload, but only stores it.

In fact, with that we only have file transport verification only on *downloads*. Also the server stores checksums without verifying them against the uploaded file first.

**Comment by DeepDiver1975** [comment]
> The server checksums the file when it arrives at the server, compares it to the checksum transmitted by the client previously
> If the checksums match, a success return code is sent
> If the chechsums fail to match, the file has been corrupted and the desktop skips that file. The file will be retransmitted on the next upload.

these criteria above should have been added to the corresponding server ticket https://github.com/owncloud/core/issues/18716

@MTRichards @karlitschek @cmonteroluque this somehow got lost while coordinating client and server requirements - hopefully we can avoid this in the future

**Comment by dragotin** [comment]
for the sake of completeness, here is the current state with 9.0.

With not having the server verifying the uploaded checksum, this can happen:
        
        1. UserA uploads a file to the server with checksumA. 
        2. FIle gets corrupted during transport
        3. Server stores corrupted file and checksumA.
        4. UserB downloads the file and checksum and they don't match.
        5. UserB will invalidate the file, redownload.
        
At the same time UserA thinks his file is properly upload to the server. Which is not the case.

**Comment by mcastroSG** [comment]
Hi hi:

### Test 1

**Steps Executed**

1.-  Connect desktop client to a server that has cksum capabilities enabled
2.- Copy a big file (at least 50Mb) to Shared sync folder
3.- While syncing connect to server and modify any chunk adding some characters to be sure that final cksum and initial cksum does not match

**Results**

Right now as could be expected server does not validate cksum at uploads so in this case we could have a corrupted file updated to the server what is described right [above](https://github.com/owncloud/client/issues/3735#issuecomment-190104308) could happend.

### Test 2

**Steps Executed**

1.-  Connect desktop client to a server that has cksum capabilities enabled
2.- Add a file1 to Shared sync folder
3.- Wait till sync is done
4.- Connect to server DB and update checksum value at filecache table for the file we have uploaded at step 3
5.- Remove connection and add a new connection to the same server

**Results**

An error is raised to let us know that the file1 does not match expected cksum and sync for that file ir retried at next sync.

### Test 3

**Steps executed**

1.- Connect desktop client to a server that has cksum capabilities enabled
2.- Add a file2.eml to Shared Sync folder
3.- Wait till sync ends
4.- `touch file2.eml`
5.- Wait till sync ends again

**Results**

File2 is not synced again.

### Test 4

**Steps executed**

1.- Connect desktop client to a server that has cksum capabilities enabled
2.- Add a file2.txt to Shared Sync folder
3.- Wait till sync ends
4.- `touch file2.txt`
5.- Wait till sync ends again

**Results**

File2.txt is uploaded again.

### Test 5

**Steps executed**

1.- Connect desktop client to a server that has cksum capabilities enabled
2.- Add several files to Shared sync folder
3.- Wait till sync ends
4.- Check at DB that checksum stored matches with local checksum, i.e. `md5`

**Results**

It works fine, checksum matches.

@dragotin  what do you think we can do with this issue ¿?

**Comment by dragotin** [comment]
@mcastroSG very good work, all tests make sense and succeed. Thanks. I think you can close this.

**Comment by mcastroSG** [comment]
Thanks a lot !! 😃

**Comment by meekjt** [comment]
What is the config option for enabling support of file transport checksums on the server side? It does not seem to be documented anywhere.

**Comment by danimo** [comment]
@meekjt The option is not there yet on the server side. It will come with one of the next versions.
